### PR TITLE
FIR IDE/UAST: add compile-time constant evaluator and provider

### PR DESCRIPTION
**Pull request by jsjeon**
First 3 commits are almost identical to #3422 (i.e. need to handle some merge conflict) where an initial version of FIR-based compile-time constant evaluator was introduced and used during (annotation) serializations and exhaustiveness transformer. Simply that PR was reincarnated, and applications to FIR (serialization and resolution) were dropped. After that, following commits adjust the evaluator according to recent changes to `FirConstKind` and operators map generator.

3rd-to-last commit introduces `KtCompileTimeConstantProvider` to HL API, along with `evaluate` API.

------

While catching up changes in operators map generator, noticed the downside of map-based operator evaluation ([KT-23565](https://youtrack.jetbrains.com/issue/KT-23565)), and thus the last two commits tried to reuse FE1.0's eval utils (as per 742fef90421f98f516e28223ed99f95bf8935870)

------

Finally, while cleaning up all the aforementioned things, realized that the evaluator was designed to be used in destructive rewrite during resolution/FIR2IR, i.e., the evaluator recursively visited the sub expressions and rewrote them as a const expression if possible, and such const was reused in an inductive manner. In FIR IDE, that's not the case. For now, due to the lack of binary expression conversion in FIR UAST, things are not fully tested yet, but leaf constant and depth-1 expressions (such as 40 + 2) are handled. For more deeper inner expressions (like 4 * 10 + 2), we need to rework the evaluator; left TODOs at the moment.

### CODE REVIEW

**Comment by darthorimar** [comment]
Hi! FIR IDE soon migrating to the kotlin-ide repo. That's why generating `FirOperationsMapGenerated.kt `in fir-ide modules will not work. Could you please move it back to `:compiler:fir:evaluate` module? Sorry for inconvenience

**Comment by jsjeon** [comment]
> Hi @jsjeon! FIR IDE soon migrating to the kotlin-ide repo. That's why generating `FirOperationsMapGenerated.kt `in fir-ide modules will not work. Could you please move it back to `:compiler:fir:evaluate` module? Sorry for inconvenience

Managed to drop the commit in the middle that migrated evaluator _and_ generated map to `idea-frontned-fir`, and rebase the following.

After the last two commits, though, now generated operation maps and utils are in `frontend.common`, i.e., FIR evaluator is free from such generator (but still use generated utils). So I wonder if it's okay to migrate the evaluator to `idea-frontend-fir` again? The latest commit does so, and should be easy to drop if not desired.

**Comment by darthorimar** [comment]
> if it's okay to migrate the evaluator to `dea-frontend-fir`

Yeap, that's totally okay 👍 It's cool that you managed to reuse existing generated code :)

File: idea/idea-frontend-fir/src/org/jetbrains/kotlin/idea/fir/evaluate/FirCompileTimeConstantEvaluator.kt [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+        when (this) {
+            is ConeKotlinErrorType -> null
+            is ConeLookupTagBasedType -> {
+                when (lookupTag.name.asString()) {
+                    "Byte" -> ConstantValueKind.Byte
+                    "Double" -> ConstantValueKind.Double
+                    "Float" -> ConstantValueKind.Float
+                    "Int" -> ConstantValueKind.Int
+                    "Long" -> ConstantValueKind.Long
+                    "Short" -> ConstantValueKind.Short
```
Comment (darthorimar) [inline idea/idea-frontend-fir/src/org/jetbrains/kotlin/idea/fir/evaluate/FirCompileTimeConstantEvaluator.kt:142]: Char, String and Boolean types seems to be missing here
Also,  can be extracted to a function.

File: idea/idea-frontend-fir/src/org/jetbrains/kotlin/idea/fir/evaluate/FirCompileTimeConstantEvaluator.kt [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+
+    // TODO: better to remove `out Number` restriction
+    private fun <T : Any> T.toConstantValueKind(): ConstantValueKind<out Number>? =
+        when (this) {
+            is Byte -> ConstantValueKind.Byte
+            is Double -> ConstantValueKind.Double
+            is Float -> ConstantValueKind.Float
+            is Int -> ConstantValueKind.Int
+            is Long -> ConstantValueKind.Long
+            is Short -> ConstantValueKind.Short
```
Comment (darthorimar) [inline idea/idea-frontend-fir/src/org/jetbrains/kotlin/idea/fir/evaluate/FirCompileTimeConstantEvaluator.kt:162]: Is Char, String and Boolean missing here?

File: idea/idea-frontend-fir/src/org/jetbrains/kotlin/idea/fir/evaluate/FirCompileTimeConstantEvaluator.kt (hunk 1 above)
Comment (jsjeon) [reply idea/idea-frontend-fir/src/org/jetbrains/kotlin/idea/fir/evaluate/FirCompileTimeConstantEvaluator.kt]: Extracting it to a function sounds good. 👍 

Those missing types were due to the `out Number` restriction I left as a TODO. Just too lazy to dig further about that, and managed to handle all of them.

File: idea/idea-frontend-fir/src/org/jetbrains/kotlin/idea/fir/evaluate/FirCompileTimeConstantEvaluator.kt (hunk 2 above)
Comment (jsjeon) [reply idea/idea-frontend-fir/src/org/jetbrains/kotlin/idea/fir/evaluate/FirCompileTimeConstantEvaluator.kt:177]: Ditto. Initially, every conversion in this evaluator was limited to `Number` (I don't remember why I did that, maybe just as a starting point). Now those were converted as well.

**Review by darthorimar — APPROVED** [review]

**Pull request merged**